jax_md/elasticity.py

```diff
@@ -0,0 +1,566 @@
+    
+    def get_affine_response(strain_tensor):
+      energy_fn_general = setup_energy_fn_general(strain_tensor)
+      d2U_dRdgamma = jacfwd(jacrev(energy_fn_general,argnums=0),argnums=1)(R,0.)
+      d2U_dgamma2  = jacfwd(jacrev(energy_fn_general,argnums=1),argnums=1)(R,0.)
+      return d2U_dRdgamma, d2U_dgamma2
+
+    strain_tensors = _get_strain_tensor_list(dim, R.dtype)
+    d2U_dRdgamma_all,d2U_dgamma2_all = vmap(get_affine_response)(strain_tensors)
+
+    #Solve the system of equations.
+    energy_fn_Ronly = partial(energy_fn, **kwargs)
+    def hvp(f, primals, tangents):
+      return jvp(grad(f), primals, tangents)[1]
+    def hvp_specific_with_tether(v):
+      return hvp(energy_fn_Ronly, (R,), (v,)) + tether_strength * v
+    
+    non_affine_response_all = jsp.sparse.linalg.cg(
+        vmap(hvp_specific_with_tether), 
+        d2U_dRdgamma_all, 
+        tol=cg_tol
+        )[0]
+    #The above line should be functionally equivalent to:
+    #H0=hessian(energy_fn)(R, box=box, **kwargs).reshape(R.size,R.size) \
+    #    + tether_strength * jnp.identity(R.size)
+    #non_affine_response_all = jnp.transpose(jnp.linalg.solve(
+    #   H0, 
+    #   jnp.transpose(d2U_dRdgamma_all))
+    #   )
+
+    residual = jnp.linalg.norm(vmap(hvp_specific_with_tether)(
+        non_affine_response_all) - d2U_dRdgamma_all
+      )
+    converged = residual / jnp.linalg.norm(d2U_dRdgamma_all) < cg_tol
+
+    response_all = d2U_dgamma2_all - jnp.einsum("nij,nij->n", 
+                                                d2U_dRdgamma_all, 
+                                                non_affine_response_all)
+
+    vol_0 = quantity.volume(dim, box)
+    response_all = response_all / vol_0
+    C = _convert_responses_to_elastic_constants(response_all)
+    
+    #JAX does not allow proper runtime error handling in jitted function. 
+    # Instead, if the user requests a gradient check and the check fails,
+    # we convert C into jnp.nan's. While this doesn't raise an exception,
+    # it at least is very "loud". 
+    if gradient_check is not None:
+      maxgrad = jnp.amax(jnp.abs(grad(energy_fn)(R, **kwargs)))
+      C = lax.cond(maxgrad > gradient_check, 
+                   lambda _: jnp.nan * C, 
+                   lambda _: C, 
+                   None)
+    
+    if check_convergence: 
+      return C, converged
+    else:
+      return C
+
+  return calculate_emt
+
+
+def _get_mandel_mapping_weight(dim, dtype):
+  if dim == 2:
+    m_map  = jnp.array([[0,0],[1,1],[0,1]], dtype=jnp.int8)
+    weight = jnp.array([1,1,jnp.sqrt(2)], dtype=dtype)
+    return m_map, weight
+  elif dim == 3:
+    m_map  = jnp.array([[0,0],[1,1],[2,2],[1,2],[0,2],[0,1]], dtype=jnp.int8)
+    weight = jnp.array([1,1,1,jnp.sqrt(2),jnp.sqrt(2),jnp.sqrt(2)], dtype=dtype)
+    return m_map, weight
+  else:
+    raise AssertionError('dim must be 2 or 3')
+
+def tensor_to_mandel(T: Array) -> Array:
+  """ Convert a tensor to Mandel notation
```

Comment:
Add a period.

Reply:
done. Found some other places where this was also needed.